stdlib/http/src/main/java/org/ballerinalang/net/http/websocket/WebSocketUtil.java

```diff
@@ -217,11 +237,459 @@ public static ErrorValue createErrorCause(String message, String reason, BPackag
+        return false;
+    }
+
+    /**
+     * Set waiting time before attempting to next doReconnect/failover.
+     * @param interval interval
+     */
+    private static void setCountDownLatch(int interval) {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+        if (wait(countDownLatch, interval)) {
+            countDownLatch.countDown();
+        }
+    }
+
+    public static boolean hasRetryConfig(ObjectValue webSocketClient) {
+        return webSocketClient.getMapValue(WebSocketConstants.CLIENT_ENDPOINT_CONFIG).
+                getMapValue(WebSocketConstants.RETRY_CONFIG) != null;
     }
 
+    private static boolean wait(CountDownLatch countDownLatch, int interval) {
+        try {
+            return countDownLatch.await(interval, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new WebSocketException("Error occurred: " + e.getMessage());
+        }
+    }
+
+    /**
+     * Calculate the waiting time.
+     * @param interval interval
+     * @param maxInterval maximum Interval
+     * @param backOfFactor back of factor
+     * @param reconnectAttempts no of doReconnect attempts
+     * @return waiting time
+     */
+    private static int getWaitTime(int interval, int maxInterval, float backOfFactor, int reconnectAttempts) {
+        interval = (int) (interval * Math.pow(backOfFactor, reconnectAttempts));
+        if (interval > maxInterval) {
+            interval = maxInterval;
+        }
+        return interval;
+    }
+
+    public static void determineAction(WebSocketOpenConnectionInfo connectionInfo, Throwable throwable,
+                                WebSocketCloseMessage webSocketCloseMessage) {
+        ObjectValue webSocketClient = connectionInfo.getWebSocketCaller();
+        if (hasRetryConfig(webSocketClient)) {
+            // When connection lost, do the failover to the remaining server urls.
+            // If failover fails, do the reconnection
+            if (failoverAndRetry(connectionInfo)) {
+                return;
+            }
+        } else {
+            // When connection lost, do the failover to the remaining server urls.
+            if (failover(connectionInfo)) {
+                return;
+            }
+        }
+        setError(connectionInfo, throwable, webSocketCloseMessage);
+    }
+
+    public static void handleExceptionAndDispatchCloseMessage(WebSocketOpenConnectionInfo connectionInfo,
+                                                       WebSocketCloseMessage webSocketCloseMessage) {
+        try {
+            WebSocketResourceDispatcher.dispatchOnClose(connectionInfo, webSocketCloseMessage);
+        } catch (IllegalAccessException e) {
+            // Ignore as it is not possible have an Illegal access
+        }
+    }
+
+    private static void closeConnection(WebSocketOpenConnectionInfo connectionInfo, Throwable throwable,
```

Comment:
I don't see this function closing any connection. It seems to be dispatching exceptions or close message. Shall we have a proper name?